src/api/c/binary.cpp

```diff
@@ -46,6 +48,96 @@ static inline af_array arithOp(const af_array lhs, const af_array rhs,
+    const dim4 ldims = linfo.dims();
+    const dim4 rdims = rinfo.dims();
+
+    unsigned maxndims =
+        ldims.ndims() > rdims.ndims() ? ldims.ndims() : rdims.ndims();
+
+    // find how much each dim4 needs to be padded
+    unsigned loffset = maxndims - ldims.ndims();
+    unsigned roffset = maxndims - rdims.ndims();
+
+    // pre-pad each dimension with 1s if necessary
+    dim4 ldims_padded = paddims(ldims, loffset);
+    dim4 rdims_padded = paddims(rdims, roffset);
+
+    dim4 bdims;
+    dim4 ltile(1, 1, 1, 1);
+    dim4 rtile(1, 1, 1, 1);
+    // check that dimensions are compatible
+    for (unsigned i = 0; i < 4; ++i) {
+        if (ldims_padded[i] == rdims_padded[i]) {
+            bdims[i] = ldims_padded[i];
+        } else if (ldims_padded[i] == 1 && rdims_padded[i] != 1) {
+            bdims[i] = rdims_padded[i];
+            ltile[i] = rdims_padded[i];
+        } else if (ldims_padded[i] != 1 && rdims_padded[i] == 1) {
+            bdims[i] = ldims_padded[i];
+            rtile[i] = ldims_padded[i];
+        } else {
+            // if both are vectors, no matching broadcasts are possible
+            if (linfo.isVector() && rinfo.isVector()) {
+                AF_ERROR("Cannot broadcast mismatching input dimensions",
+                         AF_ERR_SIZE);
+            } else if (linfo.isVector()) {
+                // restart broadcast search with different padding for vector
+                if (loffset--) {
+                    ldims_padded = paddims(ldims, loffset);
+                    i            = 0;
+                } else {
+                    // no posible padding can be broadcast
+                    AF_ERROR(
+                        "Could not find valid broadcast between vector and array,\
+                              mismatching input dimensions",
+                        AF_ERR_SIZE);
+                }
+            } else if (rinfo.isVector()) {
+                if (roffset--) {
+                    rdims_padded = paddims(rdims, roffset);
+                    i            = 0;
+                } else {
+                    AF_ERROR(
+                        "Could not find valid broadcast between vector and array,\
+                              mismatching input dimensions",
+                        AF_ERR_SIZE);
+                }
+            }
+        }
+    }
+    return std::make_tuple(bdims, ltile, rtile, ldims_padded, rdims_padded);
+}
+
+template<typename T, af_op_t op>
+static inline af_array arithOpBroadcast(const af_array lhs,
+                                        const af_array rhs) {
+    const ArrayInfo &linfo = getInfo(lhs);
+    const ArrayInfo &rinfo = getInfo(rhs);
+
+    dim4 odims, ltile, rtile, lpad, rpad;
+    std::tie(odims, ltile, rtile, lpad, rpad) = getBroadcastDims(linfo, rinfo);
+
+    detail::Array<T> lhst =
+        detail::tile<T>(modDims(getArray<T>(lhs), lpad), ltile);
+    detail::Array<T> rhst =
+        detail::tile<T>(modDims(getArray<T>(rhs), rpad), rtile);
```

Comment:
Add using statements for the following to fit convention we are following in other files.
- `using detail::Array`
- `using detail::tile`
- `using std::tie`